readme.md
This is a for from modesty and I'm going to extend the functionality for my needs.


# pdf2json

pdf2json is a [node.js](http://nodejs.org/) module that parses and converts PDF from binary to json format, it's built with [pdf.js](https://github.com/mozilla/pdf.js/) and extends it with interactive form elements and text content parsing outside browser.

The goal is to enable server side PDF parsing with interactive form elements when wrapped in web service, and also enable parsing local PDF to json file when using as a command line utility.

## Install

>npm install pdf2json

Or, install it globally:
>sudo npm install pdf2json -g

To update with latest version:
>sudo npm update pdf2json -g

To Run in RESTful Web Service or as Commandline Utility
* More details can be found at the bottom of this document.

## Code Example

* Parse a PDF file then write to a JSON file:

````javascript
    let fs = require('fs'),
        PDFParser = require("pdf2json");

    let pdfParser = new PDFParser();

    pdfParser.on("pdfParser_dataError", errData => console.error(errData.parserError) );
    pdfParser.on("pdfParser_dataReady", pdfData => {
        fs.writeFile("./pdf2json/test/F1040EZ.json", JSON.stringify(pdfData));
    });

    pdfParser.loadPDF("./pdf2json/test/pdf/fd/form/F1040EZ.pdf");
````

Or, call directly with buffer:

````javascript
    fs.readFile(pdfFilePath, (err, pdfBuffer) => {
      if (!err) {
        pdfParser.parseBuffer(pdfBuffer);
      }
    })
````

* Parse a PDF then write a .txt file (which only contains textual content of the PDF)

````javascript
    let fs = require('fs'),
        PDFParser = require("pdf2json");

    let pdfParser = new PDFParser(this,1);

    pdfParser.on("pdfParser_dataError", errData => console.error(errData.parserError) );
    pdfParser.on("pdfParser_dataReady", pdfData => {
        fs.writeFile("./pdf2json/test/F1040EZ.content.txt", pdfParser.getRawTextContent());
    });

    pdfParser.loadPDF("./pdf2json/test/pdf/fd/form/F1040EZ.pdf");
````

* Parse a PDF then write a fields.json file that only contains interactive forms' fields information:

````javascript
    let fs = require('fs'),
        PDFParser = require("pdf2json");

    let pdfParser = new PDFParser();

    pdfParser.on("pdfParser_dataError", errData => console.error(errData.parserError) );
    pdfParser.on("pdfParser_dataReady", pdfData => {
        fs.writeFile("./pdf2json/test/F1040EZ.fields.json", JSON.stringify(pdfParser.getAllFieldsTypes()));
    });

    pdfParser.loadPDF("./pdf2json/test/pdf/fd/form/F1040EZ.pdf");
````

Alternatively, you can pipe input and output streams: (requires v1.1.4)

````javascript
    let fs = require('fs'),
        PDFParser = require("pdf2json");
    
    let inputStream = fs.createReadStream("./pdf2json/test/pdf/fd/form/F1040EZ.pdf", {bufferSize: 64 * 1024});
    let outputStream = fs.createWriteStream("./pdf2json/test/target/fd/form/F1040EZ.json");
    
    inputStream.pipe(new PDFParser()).pipe(new StringifyStream()).pipe(outputStream);
````
See [p2jcmd.js](https://github.com/modesty/pdf2json/blob/master/lib/p2jcmd.js) for more details.

 
## API Reference

* events:
    * pdfParser_dataError: will be raised when parsing failed
    * pdfParser_dataReady: when parsing succeeded

* start to parse PDF file from specified file path asynchronously:
````javascript
        function loadPDF(pdfFilePath);
````
If failed, event "pdfParser_dataError" will be raised with error object: {"parserError": errObj};
If success, event "pdfParser_dataReady" will be raised with output data object: {"formImage": parseOutput}, which can be saved as json file (in command line) or serialized to json when running in web service.

* Get all textual content from "pdfParser_dataReady" event handler:
````javascript
        function getRawTextContent();
````
returns text in string.

* Get all input fields information from "pdfParser_dataReady" event handler: 
````javascript
        function getAllFieldsTypes();
````        
returns an array of field objects.         

## Output format Reference

Current parsed data has four main sub objects to describe the PDF document.

* 'Agency': the main text identifier for the PDF document. If Id.AgencyId present, it'll be same, otherwise it'll be set as document title;
* 'Transcoder': pdf2json version number
* 'Id': the XML meta data that embedded in PDF document
    * all forms attributes metadata are defined in "Custom" tab of "Document Properties" dialog in Acrobat Pro;
    * v0.1.22 added support for the following custom properties:
        * AgencyId: default "unknown";
        * Name: default "unknown";
        * MC: default false;
        * Max: default -1;
        * Parent: parent name, default "unknown";
* 'Pages': array of 'Page' object that describes each page in the PDF, including sizes, lines, fills and texts within the page. More info about 'Page' object can be found at 'Page Object Reference' section
* 'Width': the PDF page width in page unit

### Page object Reference

Each page object within 'Pages' array describes page elements and attributes with 5 main fields:

* 'Height': height of the page in page unit
* 'HLines': horizontal line array, each line has 'x', 'y' in relative coordinates for positioning, and 'w' for width, plus 'l' for length. Both width and length are in page unit
* 'Vline': vertical line array, each line has 'x', 'y' in relative coordinates for positioning, and 'w' for width, plus 'l' for length. Both width and length are in page unit;
    * v0.4.3 added Line color support. Default is 'black', other wise set in 'clr' if found in color dictionary, or 'oc' field if not found in dictionary;
    * v0.4.4 added dashed line support. Default is 'solid', if line style is dashed line, {dsh:1} is added to line object;
* 'Fills': an array of rectangular area with solid color fills, same as lines, each 'fill' object has 'x', 'y' in relative coordinates for positioning, 'w' and 'h' for width and height in page unit, plus 'clr' to reference a color with index in color dictionary. More info about 'color dictionary' can be found at 'Dictionary Reference' section.
* 'Texts': an array of text blocks with position, actual text and styling information:
    * 'x' and 'y': relative coordinates for positioning
    * 'clr': a color index in color dictionary, same 'clr' field as in 'Fill' object. If a color can be found in color dictionary, 'oc' field will be added to the field as 'original color" value.
    * 'A': text alignment, including:
        * left
        * center
        * right
    * 'R': an array of text run, each text run object has two main fields:
        * 'T': actual text
        * 'S': style index from style dictionary. More info about 'Style Dictionary' can be found at 'Dictionary Reference' section
        * 'TS': [fontFaceId, fontSize, 1/0 for bold, 1/0 for italic]

v0.4.5 added support when fields attributes information is defined in external xml file. pdf2json will always try load field attributes xml file based on file name convention (pdfFileName.pdf's field XML file must be named pdfFileName_fieldInfo.xml in the same directory). If found, fields info will be injected.

### Dictionary Reference

Same reason to having "HLines" and "VLines" array in 'Page' object, color and style dictionary will help to reduce the size of payload when transporting the parsing object over the wire.
This dictionary data contract design will allow the output just reference a dictionary key , rather than the actual full definition of color or font style.
It does require the client of the payload to have the same dictionary definition to make sense out of it when render the parser output on to screen.

* Color Dictionary

        var kColors = [
                '#000000',		// 0
                '#ffffff',		// 1
                '#4c4c4c',		// 2
                '#808080',		// 3
                '#999999',		// 4
                '#c0c0c0',		// 5
                '#cccccc',		// 6
                '#e5e5e5',		// 7
                '#f2f2f2',		// 8
                '#008000',		// 9
                '#00ff00',		// 10
                '#bfffa0',		// 11
                '#ffd629',		// 12
                '#ff99cc',		// 13
                '#004080',		// 14
                '#9fc0e1',		// 15
                '#5580ff',		// 16
                '#a9c9fa',		// 17
                '#ff0080',		// 18
                '#800080',		// 19
                '#ffbfff',		// 20
                '#e45b21',		// 21
                '#ffbfaa',		// 22
                '#008080',		// 23
                '#ff0000',		// 24
                '#fdc59f',		// 25
                '#808000',		// 26
                '#bfbf00',		// 27
                '#824100',		// 28
                '#007256',		// 29
                '#008000',		// 30
                '#000080',		// Last + 1
                '#008080',		// Last + 2
                '#800080',		// Last + 3
                '#ff0000',		// Last + 4
                '#0000ff',		// Last + 5
                '#008000',		// Last + 6
                '#000000'		// Last + 7
            ];


* Style Dictionary:

            var _kFontFaces = [
               "QuickType,Arial,Helvetica,sans-serif",							// 00 - QuickType - sans-serif variable font
               "QuickType Condensed,Arial Narrow,Arial,Helvetica,sans-serif",	// 01 - QuickType Condensed - thin sans-serif variable font
               "QuickTypePi",													// 02 - QuickType Pi
               "QuickType Mono,Courier New,Courier,monospace",					// 03 - QuickType Mono - san-serif fixed font
               "OCR-A,Courier New,Courier,monospace",							// 04 - OCR-A - OCR readable san-serif fixed font
               "OCR B MT,Courier New,Courier,monospace"							// 05 - OCR-B MT - OCR readable san-serif fixed font
            ];

            var _kFontStyles = [
                // Face		Size	Bold	Italic		StyleID(Comment)
                // -----	----	----	-----		-----------------
                    [0,		6,		0,		0],			//00
                    [0,		8,		0,		0],			//01
                    [0,		10,		0,		0],			//02
                    [0,		12,		0,		0],			//03
                    [0,		14,		0,		0],			//04
                    [0,		18,		0,		0],			//05
                    [0,		6,		1,		0],			//06
                    [0,		8,		1,		0],			//07
                    [0,		10,		1,		0],			//08
                    [0,		12,		1,		0],			//09
                    [0,		14,		1,		0],			//10
                    [0,		18,		1,		0],			//11
                    [0,		6,		0,		1],			//12
                    [0,		8,		0,		1],			//13
                    [0,		10,		0,		1],			//14
                    [0,		12,		0,		1],			//15
                    [0,		14,		0,		1],			//16
                    [0,		18,		0,		1],			//17
                    [0,		6,		1,		1],			//18
                    [0,		8,		1,		1],			//19
                    [0,		10,		1,		1],			//20
                    [0,		12,		1,		1],			//21
                    [0,		14,		1,		1],			//22
                    [0,		18,		1,		1],			//23
                    [1,		6,		0,		0],			//24
                    [1,		8,		0,		0],			//25
                    [1,		10,		0,		0],			//26
                    [1,		12,		0,		0],			//27
                    [1,		14,		0,		0],			//28
                    [1,		18,		0,		0],			//29
                    [1,		6,		1,		0],			//30
                    [1,		8,		1,		0],			//31
                    [1,		10,		1,		0],			//32
                    [1,		12,		1,		0],			//33
                    [1,		14,		1,		0],			//34
                    [1,		18,		1,		0],			//35
                    [1,		6,		0,		1],			//36
                    [1,		8,		0,		1],			//37
                    [1,		10,		0,		1],			//38
                    [1,		12,		0,		1],			//39
                    [1,		14,		0,		1],			//40
                    [1,		18,		0,		1],			//41
                    [2,		8,		0,		0],			//42
                    [2,		10,		0,		0],			//43
                    [2,		12,		0,		0],			//44
                    [2,		14,		0,		0],			//45
                    [2,		12,		0,		0],			//46
                    [3,		8,		0,		0],			//47
                    [3,		10,		0,		0],			//48
                    [3,		12,		0,		0],			//49
                    [4,		12,		0,		0],			//50
                    [0,		9,		0,		0],			//51
                    [0,		9,		1,		0],			//52
                    [0,		9,		0,		1],			//53
                    [0,		9,		1,		1],			//54
                    [1,		9,		0,		0],			//55
                    [1,		9,		1,		0],			//56
                    [1,		9,		1,		1],			//57
                    [4,		10,		0,		0],			//58
                    [5,		10,		0,		0],			//59
                    [5,		12,		0,		0]			//60
            ];


## Interactive Forms Elements

v0.1.5 added interactive forms element parsing, including text input, radio button, check box, link button and drop down list.

Interactive forms can be created and edited in Acrobat Pro for AcroForm, or in LiveCycle Designer ES for XFA forms. Current implementation for buttons only supports "link button": when clicked, it'll launch a URL specified in button properties. Examples can be found at f1040ezt.pdf file under test/data folder.

All interactive form elements parsing output will be part of corresponding 'Page' object where they belong to, radio buttons and check boxes are in 'Boxsets' array while all other elements objects are part of 'Fields' array.

Each object with in 'Boxset' can be either checkbox or radio button, the only difference is that radio button object will have more than one element in 'boxes' array, it indicates it's a radio button group. The following sample output illustrate one checkbox ( Id: F8888 ) and one radio button group ( Id: ACC ) in the 'Boxsets' array:

             Boxsets: [
                {//first element, check box
                    boxes: [ //only one box object object in this array
                    {
                        x: 47,
                        y: 40,
                        w: 3,
                        h: 1,
                        style: 48,
                        TI: 39,
                        AM: 4,
                        id: {
                            Id: "F8888",
                        },
                        T: {
                            Name: "box"
                        }
                     }
                     ],
                     id: {
                        Id: "A446",
                     }
                },//end of first element
                {//second element, radio button group
                    boxes: [// has two box elements in boxes array
                    {
                        x: 54,
                        y: 41,
                        w: 3,
                        h: 1,
                        style: 48,
                        TI: 43,
                        AM: 132,
                        id: {
                            Id: "ACCC",
                        },
                        T: {
                            Name: "box"
                        }
                    },
                    {
                        x: 67,
                        y: 41,
                        w: 3,
                        h: 1,
                        style: 48,
                        TI: 44,
                        AM: 132,
                        id: {
                            Id: "ACCS",
                            EN: 0
                        },
                        T: {
                            Name: "box"
                        }
                    }
                    ],
                    id: {
                        Id: "ACC",
                        EN: 0
                    }
                }//end of second element
             ] //end of Boxsets array

'Fields' array contains parsed object for text input (Name: 'alpha'), drop down list (Name: 'apha', but has 'PL' object which contains label array in 'PL.D' and value array in 'PL.V'), link button (Name: 'link', linked URL is in 'FL.form.Id' field). Some examples:

Text input box example:

                {
                    style: 48,
                    T: {
                        Name: "alpha",
                        TypeInfo: { }
                    },
                    id: {
                        Id: "p1_t40",
                        EN: 0
                    },
                    TU: "alternative text", //for accessibility, added only when available from PDF stream. (v0.3.6).
                    TI: 0,
                    x: 6.19,
                    y: 5.15,
                    w: 30.94,
                    h: 0.85,
                    V: "field value" //only available when the text input box has value
                },

Note: v0.7.0 extends TU (Alternative Text) to all interactive fields to better support accessibility.

Drop down list box example:

               {
                    x: 60,
                    y: 11,
                    w: 4,
                    h: 1,
                    style: 48,
                    TI: 13,
                    AM: 388,
                    mxL: 2,
                    id: {
                        Id: "ST",
                        EN: 0
                    },
                    T: {
                        Name: "alpha",
                        TypeInfo: {
                        }
                    },
                    PL: {
                        V: [
                            "",
                            "AL",
                            "AK"
                        ],
                        D: [
                        "%28no%20entry%29",
                        "Alabama",
                        "Alaska"
                        ]
                    }
               }


Link button example:

                {
                    style: 48,
                    T: {
                        Name: "link"
                    },
                    FL: {form: {Id:"http://www.github.com"},
                    id: {
                        Id: "quad8",
                        EN: 0
                    },
                    TI: 0,
                    x: 52.35,
                    y: 28.35,
                    w: 8.88,
                    h: 0.85
                }

v0.2.2 added support for "field attribute mask", it'd be common for all fields, form author can set it in Acrobat Pro's Form Editing mode: if a field is ReadOnly, it's AM field will be set as 0x00000400, otherwise AM will be set as 0.

Another supported field attributes is "required": when form author mark a field is "required" in Acrobat, the parsing result for 'AM' will be set as 0x00000010.

"Read-Only" filed attribute mask example:

                {
                    style: 48,
                    T: {
                        Name: "alpha",
                        TypeInfo: { }
                    },
                    id: {
                        Id: "p1_t40",
                        EN: 0
                    },
                    TI: 0,
                    AM: 1024, //If (AM & 0x00000400) set, it indicates this is a read-only filed
                    x: 6.19,
                    y: 5.15,
                    w: 30.94,
                    h: 0.85
                }

## Text Input Field Formatter Types

v0.1.8 added text input field formatter types detection for

* number
* ssn
* date (tested date formatter: mm/dd/yyyy, mm/dd, mm/yyyy and Custom yyyy)
* zip
* phone
* percent (added v0.5.6)

v0.3.9 added "arbitrary mask" (in "special" format category) support, the input field format type is "mask" and the mask string is added as "MV", its value can be found at Format => Special => Arbitrary Mask in Acrobat;
Some examples of "mask" format including:
* 9999: 4 digit PIN field
* 99999: 5 digit PIN field
* 99-9999999: formatted 9 digit EIN number
* 999999999: 9 digit routing number
* aaa: 3 letters input

Additionally, the "arbitrary mask" length is extended from 1 characters to 64 characters. And when the mask has only one character, it has the following meanings:
* a: alphabet only input, no numeric input allowed
* n: numeric only input, no locale based number formatting, no alphabet or special characters allowed
* d: numeric only input, with locale based number formatting, one decimal point allowed, no rounding expected and no alphabet or special characters allowed
* -: negative number only, with locale based number formatting, no alphabet or special characters allowed
* +: positive number only, with locale based number formatting, no alphabet or special characters allowed

v0.4.1 added more date format detection, these formats are set in Acrobat's field's Properties => Format => Date => Custom:
* yyyy: 4 digit year

Types above are detected only when the widget field type is "Tx" and the additional-actions dictionary 'AA' is set. Like what you see, not all pre-defined formatters and special formatters are supported, if you need more support, you can extend the 'processFieldAttribute' function in core.js file.

For the supported types, the result data is set to the field item's T object. Example of a 'number' field in final json output:

                {
                    style: 48,
                    T: {
                        Name: "number",
                        TypeInfo: { }
                    },
                    id: {
                        Id: "FAGI",
                        EN: 0
                    },
                    TI: 0,
                    x: 68.35,
                    y: 22.43,
                    w: 21.77,
                    h: 1.08
                },

Another example of 'date' field:

                {
                    style: 48,
                    T: {
                        Name: "date",
                        TypeInfo: { }
                    },
                    id: {
                        Id: "Your Birth Date",
                        EN: 0
                    },
                    TI: 0,
                    x: 33.43,
                    y: 20.78,
                    w: 5.99,
                    h: 0.89
                },


## Text Style data without Style Dictionary

v0.1.11 added text style information in addition to style dictionary. As we discussed earlier, the idea of style dictionary is to make the parsing result payload to be compact, but I found out the limited dictionary entries for font (face, size) and style (bold, italic) can not cover majority of text contents in PDFs, because of some styles are matched with closest dictionary entry, the client rendering will have mis-aligned, gapped or overlapped text. To solve this problem, pdf2json v0.1.11 extends the dictionary approach, all previous dictionary entries stay the same, but parsing result will not try to match to a closest style entry, instead, all exact text style will be returned in a TS filed.

When the actual text style doesn't match any pre-defined style dictionary entry, the text style ID (S filed) will be set as -1. The actual text style will be set in a new field (TS) with or without a matched style dictionary entry ID. This means, if your client renderer works with pdf2json v0.1.11 and later, style dictionary ID can be ignored. Otherwise, previous client renderer can still work with style dictionary ID.

The new TS filed is an Array with format as:

* First element in TS Array is Font Face ID (integer)
* Second element is Font Size (px)
* Third is 1 when font weight is bold, otherwise 0
* Forth is 1 when font style is italic, otherwise 0

For example, the following is a text block data in the parsing result:

                {
                    x: 7.11,
                    y: 2.47,
                    w: 1.6,
                    clr: 0,
                    A: "left",
                    R: [
                        {
                            T: "Modesty%20PDF%20Parser%20NodeJS",
                            S: -1,
                            TS: [0, 15, 1, 0]
                        }
                    ]
                },

The text is "Modesty PDF Parser NodeJS", text style dictionary entry ID is -1 (S field, meaning no match), and its Font Face ID is 0 (TS[0], "QuickType,Arial,Helvetica,sans-serif"), Font Size is 15px (TS[1]), Font weight is bold (TS[2]) and font style is normal (TS[3]).


History.md

1.1.10 / 2019-09-18
==================

**fixes**
  * [[`1f52cc5`](http://github.com/yuque/pdf2json/commit/1f52cc532e19ede03d27f0c95d28af536d1215c0)] - fix: make sure annotation MK exists (#1) (fengmk2 <<fengmk2@gmail.com>>)

**others**
  * [[`c833ecb`](http://github.com/yuque/pdf2json/commit/c833ecb20ab58e12579be7975ca394fa83884469)] - Update lodash to 4.17.15 (drabinowitz <<dmitrirabinowitz+code@gmail.com>>)

1.1.9 / 2019-07-17
==================

**fixes**
  * [[`274ac6e`](http://github.com/yuque/pdf2json/commit/274ac6e5a124f15528613a9f9faa5c99139a4c21)] - fix: detect this[fnId] exists or not (fengmk2 <<fengmk2@gmail.com>>)

1.1.8 / 2019-06-27
==================

**fixes**
  * [[`15ef86a`](http://github.com/yuque/pdf2json/commit/15ef86a6c1362d3de8572259e8b787eda54e26df)] - fix: allow undefined value on XRef_getEntry (fengmk2 <<fengmk2@gmail.com>>),


pdffont.js
var nodeUtil = require("util"),
    _ = require("underscore"),
    PDFUnit = require('./pdfunit.js');

var PDFFont = (function PFPFontClosure() {
    'use strict';
    // private static
    var _nextId = 1;
    var _name = 'PDFFont';

    var _boldSubNames = ["bd", "bold", "demi", "black"];

    var _kFontFaces = [
       "QuickType,Arial,Helvetica,sans-serif",							// 00 - QuickType - sans-serif variable font
       "QuickType Condensed,Arial Narrow,Arial,Helvetica,sans-serif",	// 01 - QuickType Condensed - thin sans-serif variable font
       "QuickTypePi",													// 02 - QuickType Pi
       "QuickType Mono,Courier New,Courier,monospace",					// 03 - QuickType Mono - san-serif fixed font
       "OCR-A,Courier New,Courier,monospace",							// 04 - OCR-A - OCR readable san-serif fixed font
       "OCR B MT,Courier New,Courier,monospace"							// 05 - OCR-B MT - OCR readable san-serif fixed font
    ];

    var _kFontStyles = [
        // Face		Size	Bold	Italic		StyleID(Comment)
        // -----	----	----	-----		-----------------
            [0,		6,		0,		0],			//00
            [0,		8,		0,		0],			//01
            [0,		10,		0,		0],			//02
            [0,		12,		0,		0],			//03
            [0,		14,		0,		0],			//04
            [0,		18,		0,		0],			//05
            [0,		6,		1,		0],			//06
            [0,		8,		1,		0],			//07
            [0,		10,		1,		0],			//08
            [0,		12,		1,		0],			//09
            [0,		14,		1,		0],			//10
            [0,		18,		1,		0],			//11
            [0,		6,		0,		1],			//12
            [0,		8,		0,		1],			//13
            [0,		10,		0,		1],			//14
            [0,		12,		0,		1],			//15
            [0,		14,		0,		1],			//16
            [0,		18,		0,		1],			//17
            [0,		6,		1,		1],			//18
            [0,		8,		1,		1],			//19
            [0,		10,		1,		1],			//20
            [0,		12,		1,		1],			//21
            [0,		14,		1,		1],			//22
            [0,		18,		1,		1],			//23
            [1,		6,		0,		0],			//24
            [1,		8,		0,		0],			//25
            [1,		10,		0,		0],			//26
            [1,		12,		0,		0],			//27
            [1,		14,		0,		0],			//28
            [1,		18,		0,		0],			//29
            [1,		6,		1,		0],			//30
            [1,		8,		1,		0],			//31
            [1,		10,		1,		0],			//32
            [1,		12,		1,		0],			//33
            [1,		14,		1,		0],			//34
            [1,		18,		1,		0],			//35
            [1,		6,		0,		1],			//36
            [1,		8,		0,		1],			//37
            [1,		10,		0,		1],			//38
            [1,		12,		0,		1],			//39
            [1,		14,		0,		1],			//40
            [1,		18,		0,		1],			//41
            [2,		8,		0,		0],			//42
            [2,		10,		0,		0],			//43
            [2,		12,		0,		0],			//44
            [2,		14,		0,		0],			//45
            [2,		12,		0,		0],			//46 MQZ: Changed font size from 18 tp 12
            [3,		8,		0,		0],			//47
            [3,		10,		0,		0],			//48
            [3,		12,		0,		0],			//49
            [4,		12,		0,		0],			//50
            [0,		9,		0,		0],			//51
            [0,		9,		1,		0],			//52
            [0,		9,		0,		1],			//53
            [0,		9,		1,		1],			//54
            [1,		9,		0,		0],			//55
            [1,		9,		1,		0],			//56
            [1,		9,		1,		1],			//57
            [4,		10,		0,		0],			//58
            [5,		10,		0,		0],			//59
            [5,		12,		0,		0]			//60
    ];


    // constructor
    var cls = function (fontObj) {
        // private
        var _id = _nextId++;

        // public (every instance will have their own copy of these methods, needs to be lightweight)
        this.get_id = function() { return _id; };
        this.get_name = function() { return _name + _id; };

        this.fontObj = fontObj;
        var typeName = (fontObj.name || fontObj.fallbackName).toLowerCase();
        this.isSymbol = typeName.indexOf("symbol") > 0;
        if (this.fontObj.isSymbolicFont) {
            if (typeName.indexOf("arial") > 0)
                this.fontObj.isSymbolicFont = false; //lots of Arial-based font is detected as symbol in VA forms (301, 76-c, etc.) reset the flag for now
        }
        else {
            if (this.isSymbol)
                this.fontObj.isSymbolicFont = true; //text pdf: va_ind_760c
        }

        this.fontSize = 1;
        this.typeName = typeName;

        this.faceIdx = 0;
        this.bold = false;
        this.italic = false;

        this.fontStyleId = -1;
    };

    // public static
    cls.get_nextId = function () {
        return _name + _nextId;
    };

    // private
    var _setFaceIndex = function() {
        var fontObj = this.fontObj;

        this.bold = fontObj.bold;
        if (!this.bold) {
            this.bold = this.typeName.indexOf("bold") >= 0 || this.typeName.indexOf("black") >= 0;
        }

        var typeName = "";
        var nameArray = this.typeName.split('+');
        if (_.isArray(nameArray) && nameArray.length > 1) {
            typeName = nameArray[1].split("-");
            if (_.isArray(typeName) && typeName.length > 1) {
                if (!this.bold) {
                    var subName = typeName[1].toLowerCase();
                    this.bold = _boldSubNames.indexOf(subName) >= 0;
                }
                typeName = typeName[0];
            }
        }

        if (fontObj.isSerifFont) {
            if (_kFontFaces[1].indexOf(typeName) >= 0)
                this.faceIdx = 1;
        }
        else if (fontObj.isMonospace) {
            this.faceIdx = 3;

            if (_kFontFaces[4].indexOf(typeName) >= 0)
                this.faceIdx = 4;
            else if (_kFontFaces[5].indexOf(typeName) >= 0)
                this.faceIdx = 5;
        }
        else if (fontObj.isSymbolicFont) {
            this.faceIdx = 2;
        }

        if (this.faceIdx == 0) {
            if (this.typeName.indexOf("narrow") > 0)
                this.faceIdx = 1;
        }

//        nodeUtil._logN.call(this, "typeName = " + typeName + " => faceIdx = " + this.faceIdx);
    };

    var _getFontStyleIndex = function(fontSize) {
        _setFaceIndex.call(this);

        //MQZ Feb.28.2013. Adjust bold text fontsize to work around word spacing issue
        this.fontSize = (this.bold && (fontSize > 12)) ? fontSize + 1 : fontSize;

        var fsa = [this.faceIdx, this.fontSize, this.bold?1:0, this.italic?1:0];
        var retVal = -1;

        _.each(_kFontStyles, function(element, index, list){
            if (retVal === -1) {
                if (element[0] === fsa[0] && element[1] === fsa[1] &&
                    element[2] === fsa[2] && element[3] === fsa[3]) {
                        retVal = index;
                }
            }
        });

        return retVal;
    };

    var _processSymbolicFont = function(str) {
        var retVal = str;

        if (!str || str.length !== 1)
            return retVal;

        if (!this.fontObj.isSymbolicFont)
            return retVal;

        switch(str.charCodeAt(0)) {
            case 99: retVal = this.isSymbol ? '\u2022' : '\u25b2'; break; //up triangle. set to Bullet Dot for VA SchSCR
            case 97: retVal = '\u25b6'; break; //right triangle
            case 20: retVal = '\u2713'; break; //check mark
            case 70: retVal = '\u007D'; break; //right curly bracket
            case 118: retVal = '\u2022'; break; //Bullet dot
            case 106: retVal = ''; break; //VA 301: string j character by the checkbox, hide it for now
            default:
                nodeUtil._logN.call(this, "Default - SymbolicFont - (" + this.fontObj.name + ") : " +
                    str.charCodeAt(0) + "::" + str.charCodeAt(1) + " => " + retVal + " length = " + str.length);
        }

        return retVal;
    };

    var _textRotationAngle = function (matrix2D) {
        var retVal = 0;
        if (matrix2D[0][0] === 0 && matrix2D[1][1] === 0) {
            if (matrix2D[0][1] != 0 && matrix2D[1][0] != 0) {
                if ((matrix2D[0][1] / matrix2D[1][0]) + 1 < 0.0001)
                    retVal = 90;
            }
        }
        else if (matrix2D[0][0] !== 0 && matrix2D[1][1] !== 0) {
            var r1 = Math.atan(-matrix2D[0][1] / matrix2D[0][0]);
            var r2 = Math.atan(matrix2D[1][0] / matrix2D[1][1]);
            if (Math.abs(r1) > 0.0001 && (r1 - r2 < 0.0001)) {
                retVal = r1 * 180 / Math.PI;
            }
        }
        return retVal;
    };

    // public (every instance will share the same method, but has no access to private fields defined in constructor)
    cls.prototype.processText = function (p, str, maxWidth, color, fontSize, targetData, matrix2D) {
//        console.log(str + ", " + JSON.stringify({x:p.x, y:p.y, width: PDFUnit.toPixelX(maxWidth)}));
        this.fontStyleId = _getFontStyleIndex.call(this, fontSize);
        var text = _processSymbolicFont.call(this, str);

        if (text == "C" || text == "G") { //prevent symbolic encoding from the client
            text = " " + text + " ";
        }

        //adjust bold big text positioning, like in va_ind_301
        if (this.bold && this.fontSize >= 20)
            p.y -= this.fontSize * 0.375;

        // when this.fontStyleId === -1, it means the text style doesn't match any entry in the dictionary
        // adding TS to better describe text style [fontFaceId, fontSize, 1/0 for bold, 1/0 for italic];
        var TS = [this.faceIdx, this.fontSize, this.bold?1:0, this.italic?1:0];

        var oneText = {x: PDFUnit.toFormX(p.x) - 0.25,
            y: PDFUnit.toFormY(p.y) - 0.75,
            w: maxWidth,
            clr: PDFUnit.findColorIndex(color),
            A: "left",
            R: [{
                T: this.flash_encode(text),
                S: this.fontStyleId,
                TS: TS
            }]
        };
        var rAngle = _textRotationAngle.call(this, matrix2D);
        if (rAngle != 0) {
            nodeUtil._logN.call(this, str + ": rotated " + rAngle + " degree.");
            _.extend(oneText.R[0], {RA: rAngle});
        }

        targetData.Texts.push(oneText);
    };

    cls.prototype.flash_encode = function(str) {
        var retVal = encodeURIComponent(str);
        retVal = retVal.replace("%C2%96", "-");
        retVal = retVal.replace("%C2%91", "%27");
        retVal = retVal.replace("%C2%92", "%27");
        retVal = retVal.replace("%C2%82", "%27");
        retVal = retVal.replace("%C2%93", "%22");
        retVal = retVal.replace("%C2%94", "%22");
        retVal = retVal.replace("%C2%84", "%22");
        retVal = retVal.replace("%C2%8B", "%C2%AB");
        retVal = retVal.replace("%C2%9B", "%C2%BB");

        return retVal;
    };

    cls.prototype.clean = function() {
        this.fontObj = null;
        delete this.fontObj;
    };

    return cls;
})();

module.exports = PDFFont;

